packages/trace-viewer/src/ui/sourceTab.tsx

```diff
@@ -43,35 +42,37 @@ export const SourceTab: React.FunctionComponent<{
   }, [action, lastAction, setLastAction, setSelectedFrame]);
 
   const { source, highlight, targetLine, fileName } = useAsyncMemo<{ source: SourceModel, targetLine?: number, fileName?: string, highlight: SourceHighlight[] }>(async () => {
-    const location = action?.stack?.[selectedFrame] || fallbackLocation;
-    if (!location?.file)
-      return { source: { errors: [], content: undefined }, targetLine: 0, highlight: [] };
+    const actionLocation = action?.stack?.[selectedFrame];
+    const shoulduseFallback = !actionLocation?.file;
+    if (shoulduseFallback && !fallbackSource)
+      return { source: { file: '', errors: [], content: undefined }, targetLine: 0, highlight: [] };
 
-    let source = sources.get(location.file);
+    const file = shoulduseFallback ? fallbackSource!.file : actionLocation.file;
+    let source = sources.get(file);
     // Fallback location can fall outside the sources model.
     if (!source) {
-      source = { errors: [], content: undefined };
-      sources.set(location.file, source);
+      source = { file, errors: fallbackSource?.errors || [], content: undefined };
+      sources.set(file, source);
     }
 
-    const targetLine = location.line || 0;
-    const fileName = rootDir && location.file.startsWith(rootDir) ? location.file.substring(rootDir.length + 1) : location.file;
-    const highlight: SourceHighlight[] = source.errors.map(e => ({ type: 'error', line: e.location.line, message: e.error!.message }));
+    const targetLine = shoulduseFallback ? fallbackSource?.errors[0]?.line || 0 : actionLocation.line;
+    const fileName = rootDir && file.startsWith(rootDir) ? file.substring(rootDir.length + 1) : file;
+    const highlight: SourceHighlight[] = source.errors.map(e => ({ type: 'error', line: e.line, message: e.message }));
     highlight.push({ line: targetLine, type: 'running' });
 
-    if (source.content === undefined || fallbackLocation) {
-      const sha1 = await calculateSha1(location.file);
+    if (source.content === undefined || fallbackSource) {
```

Comment:
I wonder why fetch when `fallbackSource` is there, but `!shoulduseFallback`?

Reply:
They are interchangeable in this context.